waiter/src/waiter/state.clj

```diff
@@ -753,6 +754,54 @@
   (log/warn "[close-update-state-channel] closing update-state-chan of" service-id)
   (async/go (async/close! update-state-chan)))
 
+(defrecord InstanceTracker [service-id known-instance-ids scheduling-instance-timers starting-instance-trackers])
+
+(defn- new-tracker-val
+  [service-id]
+  (map->InstanceTracker {:service-id service-id
+                         :known-instance-ids #{}
+                         ; TODO - figure out min instances from service-id
+                         ; TODO - track overall request-to-running time
+                         :scheduling-instance-timers [[(timers/start (metrics/waiter-timer "app-schedule-time"))
+                                                       (timers/start (metrics/service-timer service-id "app-schedule-time"))]]
+                         :starting-instance-trackers []}))
+
+(defn update-instance-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-instance-trackers new-service-ids removed-service-ids service-id->healthy-instances service-id->unhealthy-instances]
+  (log/info "Updating trackers..." [service-instance-trackers new-service-ids removed-service-ids service-id->healthy-instances service-id->unhealthy-instances])
+  (->> service-instance-trackers
+       (remove (comp removed-service-ids :service-id))
+       (concat (mapv new-tracker-val new-service-ids))
+       (mapv (fn [{:keys [service-id known-instance-ids scheduling-instance-timers starting-instance-trackers]}]
+               (log/info "Updating trackers for service" [service-id known-instance-ids scheduling-instance-timers starting-instance-trackers])
+               (let [healthy-instance-ids (->> service-id service-id->healthy-instances (map :id) set)
+                     known-instance-ids' (->> service-id service-id->unhealthy-instances (map :id) (into healthy-instance-ids))
+                     new-instance-ids (set/difference known-instance-ids' known-instance-ids)
+                     ;; Pair up all of the newly-scheduled instances with their timers.
+                     ;; NOTE: Using split-at allows us to drop the new-instances if we're somehow missing the corresponding timers
+                     [paired-timers scheduling-instance-timers'] (split-at (count new-instance-ids) scheduling-instance-timers)
```

Comment:
I think this code would be much easier to understand if the trackers were mapped against an instance id.

Reply:
We don't know the instance ID until this point. This line is what matches a request for a new instance with the instance that eventually shows up.